src/rrdp-iostat/rrdp_iostat.ml

```diff
@@ -518,19 +518,40 @@ module Stats_value = struct
       }
     | Some s3 ->
       let last_s3 = last_stats_blktap3 in
-      let opt f x = match x with None -> 0L | Some x' -> f x' in
+      let opt f = Option.fold ~none:0L ~some:f in
       let open Blktap3_stats in
       {
         rd_bytes = Int64.mul (get_stats_read_sectors s3) 512L;
         wr_bytes = Int64.mul (get_stats_write_sectors s3) 512L;
         rd_avg_usecs =
-          if (get_stats_read_reqs_completed s3) > 0L then
-            Int64.div (get_stats_read_total_ticks s3) (get_stats_read_reqs_completed s3)
-          else 0L;
-        wr_avg_usecs =
-          if (get_stats_write_reqs_completed s3) > 0L then
-            Int64.div (get_stats_write_total_ticks s3) (get_stats_write_reqs_completed s3)
-          else 0L;
+          Option.fold ~none:0L (* we can't determine num_recent_reads *) ~some:(fun last_s3 ->
+            (* a 'tick' is the time spent waiting on reads (or writes, dep on context), in usec *)
+            let num_reads_in_last_5_secs =
+              let now = get_stats_read_reqs_completed s3 in
+              let last = get_stats_read_reqs_completed last_s3 in
+              if now < last then begin
+                D.error "Stats_value.make: read_reqs_completed has decreased: now: %Li, last: %Li" now last;
+                0L
+              end else
+                now -- last
+            in
+            let ticks_in_last_5_secs = get_stats_read_total_ticks s3 -- get_stats_read_total_ticks last_s3 in
+            if num_reads_in_last_5_secs = 0L then 0L else Int64.div ticks_in_last_5_secs num_reads_in_last_5_secs (* average latency per usec *)
+          ) last_s3;
+        wr_avg_usecs = (* similar to rd_avg_usecs *)
+          Option.fold ~none:0L ~some:(fun last_s3 ->
+            let num_writes_in_last_5_secs =
+              let now = get_stats_write_reqs_completed s3 in
+              let last = get_stats_write_reqs_completed last_s3 in
+              if now <= last then begin
```

Comment:
if `now == last` that is fine, see above where you used `<`.
There is quite a lot of duplication between reads and writes here, perhaps you could have a `compute_average  get_stats_write_total_ticks get_stats_write_reqs_completed` and implement it generically so you can reuse same code for both?

Reply:
i think i meant to change this to `<` also because we don't want to log errors in the case that they are equal

@lindig  left a comment about the duplication also. i think because i made this mistake, i ought to factor this out